File: apps/macos/Sources/WispProjectBrowser/NativeContextActivity.swift
import Foundation

/// Runtime fields retain the backend's camelCase wire format.
public struct NativeRuntimeKey: Codable, Sendable {
    public let projectId: String
    public let contextId: String
    public let language: String
    public let scopeKey: String
    public let sessionId: String
}
public struct NativeRuntimeInfo: Codable, Identifiable, Sendable {
    public var id: String { runtimeId }
    public let runtimeId: String
    public let generation: UInt64
    public let key: NativeRuntimeKey
    public let status: String
    public let interpreter: String?
    public let version: String?
    public let processId: UInt32?
    public let startedAtMs: UInt64
    public let lastActivityAtMs: UInt64
    public let residentMemoryBytes: UInt64?
    public let lastError: String?
}
public struct NativeRuntimeObject: Codable, Identifiable, Sendable {
    public var id: String { name }
    public let name: String
    public let typeName: String
    public let summary: String
    public let sizeBytes: UInt64?
}
public struct NativeRuntimeObjects: Codable, Sendable {
    public let objects: [NativeRuntimeObject]
    public let totalCount: Int
}
/// Shared RunSummary/RunRecord fields. Large detail fields are absent in lists.
public struct NativeRun: Codable, Identifiable, Sendable {
    public let id: String
    public let frame_id: String?
    public let context_id: String
    public let title: String
    public let kind: String
    public let status: String
    public let created_at: Int64
    public let started_at: Int64?
    public let ended_at: Int64?
    public let exit_code: Int64?
    public let remote_workdir: String?
    public let timeout_secs: Int64?
    public let last_polled_at: Int64?
    public let last_poll_error: String?
    public let progress_json: String
    public let harvested_at: Int64?
    public let cleaned_at: Int64?
    public let cleanup_error: String?
    public let output_fingerprint: String?
    public let command: String?
    public let stdout_tail: String?
    public let stderr_tail: String?
    public let env_snapshot_json: String?
    public var cancellable: Bool { ["submitted", "running", "cancelling"].contains(status) }
    public var harvestable: Bool { status == "succeeded" && harvested_at == nil }
}
public struct NativeContextActivity: Codable, Sendable {
    public let runtimes: [NativeRuntimeInfo]
    public let runs: [NativeRun]
    public let read_only: Bool
}

public struct NativeRuntimeExecution: Codable, Sendable {
    public let text: String
    public let plots: [String]
}
public enum NativeRuntimeAction: String, Sendable {
    case stop, restart, dismiss
}

File: apps/windows/Wisp.ProjectBrowser.Contracts/NativeContextActivity.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;
namespace Wisp.ProjectBrowser.Contracts;
public sealed record NativeRuntimeKey(
    [property: JsonPropertyName("projectId")] string ProjectId,
    [property: JsonPropertyName("contextId")] string ContextId,
    string Language,
    [property: JsonPropertyName("scopeKey")] string ScopeKey,
    [property: JsonPropertyName("sessionId")] string SessionId);
public sealed record NativeRuntimeInfo(
    [property: JsonPropertyName("runtimeId")] string RuntimeId,
    ulong Generation, NativeRuntimeKey Key, string Status, string? Interpreter, string? Version,
    [property: JsonPropertyName("processId")] uint? ProcessId,
    [property: JsonPropertyName("startedAtMs")] ulong StartedAtMs,
    [property: JsonPropertyName("lastActivityAtMs")] ulong LastActivityAtMs,
    [property: JsonPropertyName("residentMemoryBytes")] ulong? ResidentMemoryBytes,
    [property: JsonPropertyName("lastError")] string? LastError);
public sealed record NativeRuntimeObject(string Name, [property: JsonPropertyName("typeName")] string TypeName, string Summary,
    [property: JsonPropertyName("sizeBytes")] ulong? SizeBytes);
public sealed record NativeRuntimeObjects(NativeRuntimeObject[] Objects, [property: JsonPropertyName("totalCount")] int TotalCount);
public sealed record NativeRun(string Id, string? FrameId, string ContextId, string Title, string Kind, string Status,
    long CreatedAt, long? StartedAt, long? EndedAt, long? ExitCode, string? RemoteWorkdir, long? TimeoutSecs,
    long? LastPolledAt, string? LastPollError, string ProgressJson, long? HarvestedAt, long? CleanedAt,
    string? CleanupError, string? OutputFingerprint, string? Command, string? StdoutTail, string? StderrTail, string? EnvSnapshotJson);
public sealed record NativeContextActivity(NativeRuntimeInfo[] Runtimes, NativeRun[] Runs, bool ReadOnly);
public interface INativeContextActivityClient
{
    Task<NativeContextActivity> ReadAsync(string project, string session, CancellationToken token = default);
    Task<NativeRun> ReadRunAsync(string project, string session, string runId, CancellationToken token = default);
    Task<NativeRuntimeObjects> InspectRuntimeAsync(string project, string session, string runtimeId, CancellationToken token = default);
    Task<NativeRun> CancelRunAsync(string project, string session, string runId, CancellationToken token = default);
    Task<NativeRun> HarvestRunAsync(string project, string session, string runId, CancellationToken token = default);
}
/// Uses explicit frame scope. Mutations are never replayed after uncertain replies.
public sealed class NativeContextActivityClient(INativeSettingsClient transport) : INativeContextActivityClient
{
    private async Task<T> Call<T>(string action, string project, string session, JsonObject args, CancellationToken token) where T : class
    {
        args["session_id"] = session;
        return (await transport.InvokeAsync("native_conversation_panel_" + action, args, project, token).ConfigureAwait(false))?.Deserialize<T>(ConversationSnapshot.JsonOptions)
            ?? throw new InvalidDataException("Missing activity response");
    }
    public Task<NativeContextActivity> ReadAsync(string project, string session, CancellationToken token = default) => Call<NativeContextActivity>("activity", project, session, new(), token);
    public Task<NativeRun> ReadRunAsync(string project, string session, string runId, CancellationToken token = default) => Call<NativeRun>("run_detail", project, session, new() { ["run_id"] = runId }, token);
    public Task<NativeRuntimeObjects> InspectRuntimeAsync(string project, string session, string runtimeId, CancellationToken token = default) => Call<NativeRuntimeObjects>("runtime_inspect", project, session, new() { ["runtime_id"] = runtimeId }, token);
    public Task<NativeRun> CancelRunAsync(string project, string session, string runId, CancellationToken token = default) => Call<NativeRun>("run_cancel", project, session, new() { ["run_id"] = runId }, token);
    public Task<NativeRun> HarvestRunAsync(string project, string session, string runId, CancellationToken token = default) => Call<NativeRun>("run_harvest", project, session, new() { ["run_id"] = runId }, token);
}
